### PR TITLE
Fix tests for `spark.DeltaTable` dataset

### PR DESCRIPTION
**Pull request by limdauto**
## Description

This PR fixes the failing tests for #964. It has been very tricky to debug, so I created this PR using `feature/databricks-deltatable-dataset` as base.

## Development notes

Below is the (wild) journey I took to get to the bottom of this one.

### 🧽   Cleanup the Spark test setup

The current Spark test setup is bloated and very hard to follow with too many indirections. To be completely honest, I don't think I fully understand the whole `the_real_getOrCreate` and `UseTheSparkSessionFixtureOrMock` business. It causes a lot of issues when trying to decipher what actually was breaking, so I removed them all.  Now the setup is dead simple:

* All spark tests have an autoused `spark_session` fixture by default provided in conftest.
* Tests that need specialised session, e.g. hive, can override this fixture by providing a fixture of the same name in its test module.
* No need for all the gc cleanup, JVM caches, etc. anymore.
 
Now it's all idiomatic pytest and just works.

### 🦠   First bug: broken `cacerts`

I think we were misguided by the CI logs. What we saw were truncated and only showed a rather unhelpful

```python
Exception: Java gateway process exited before sending its port number
```

Upon inspecting the full [logs](https://circleci.com/api/v1.1/project/github/quantumblacklabs/kedro/69832/output/110/0?file=true&allocation-id=61a8e1d7010e1b64d5ab7bda-0-build%2F5441F6DA) of the original failure, I found the real error log from the JVM:

```
---------------------------- Captured stderr setup -----------------------------
Ivy Default Cache set to: /home/circleci/.ivy2/cache
The jars for the packages stored in: /home/circleci/.ivy2/jars
io.delta#delta-core_2.12 added as a dependency
:: resolving dependencies :: org.apache.spark#spark-submit-parent-70a12613-636b-4941-bbe5-07e47b89107e;1.0
	confs: [default]
:: resolution report :: resolve 604ms :: artifacts dl 0ms
	:: modules in use:
	---------------------------------------------------------------------
	|                  |            modules            ||   artifacts   |
	|       conf       | number| search|dwnlded|evicted|| number|dwnlded|
	---------------------------------------------------------------------
	|      default     |   1   |   0   |   0   |   0   ||   0   |   0   |
	---------------------------------------------------------------------

:: problems summary ::
:::: WARNINGS
		module not found: io.delta#delta-core_2.12;1.0.0

	==== local-m2-cache: tried

	  file:/home/circleci/.m2/repository/io/delta/delta-core_2.12/1.0.0/delta-core_2.12-1.0.0.pom

	  -- artifact io.delta#delta-core_2.12;1.0.0!delta-core_2.12.jar:

	  file:/home/circleci/.m2/repository/io/delta/delta-core_2.12/1.0.0/delta-core_2.12-1.0.0.jar

	==== local-ivy-cache: tried

	  /home/circleci/.ivy2/local/io.delta/delta-core_2.12/1.0.0/ivys/ivy.xml

	  -- artifact io.delta#delta-core_2.12;1.0.0!delta-core_2.12.jar:

	  /home/circleci/.ivy2/local/io.delta/delta-core_2.12/1.0.0/jars/delta-core_2.12.jar

	==== central: tried

	  https://repo1.maven.org/maven2/io/delta/delta-core_2.12/1.0.0/delta-core_2.12-1.0.0.pom

	  -- artifact io.delta#delta-core_2.12;1.0.0!delta-core_2.12.jar:

	  https://repo1.maven.org/maven2/io/delta/delta-core_2.12/1.0.0/delta-core_2.12-1.0.0.jar

	==== spark-packages: tried

	  https://repos.spark-packages.org/io/delta/delta-core_2.12/1.0.0/delta-core_2.12-1.0.0.pom

	  -- artifact io.delta#delta-core_2.12;1.0.0!delta-core_2.12.jar:

	  https://repos.spark-packages.org/io/delta/delta-core_2.12/1.0.0/delta-core_2.12-1.0.0.jar

		::::::::::::::::::::::::::::::::::::::::::::::

		::          UNRESOLVED DEPENDENCIES         ::

		::::::::::::::::::::::::::::::::::::::::::::::

		:: io.delta#delta-core_2.12;1.0.0: not found

		::::::::::::::::::::::::::::::::::::::::::::::


:::: ERRORS
	Server access error at url https://repo1.maven.org/maven2/io/delta/delta-core_2.12/1.0.0/delta-core_2.12-1.0.0.pom (javax.net.ssl.SSLException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty)

	Server access error at url https://repo1.maven.org/maven2/io/delta/delta-core_2.12/1.0.0/delta-core_2.12-1.0.0.jar (javax.net.ssl.SSLException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty)

	Server access error at url https://repos.spark-packages.org/io/delta/delta-core_2.12/1.0.0/delta-core_2.12-1.0.0.pom (javax.net.ssl.SSLException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty)

	Server access error at url https://repos.spark-packages.org/io/delta/delta-core_2.12/1.0.0/delta-core_2.12-1.0.0.jar (javax.net.ssl.SSLException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty)
```

It says it can't resolved delta core JAR because of an SSL error: 

```
javax.net.ssl.SSLException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty
```

A quick Google search led me to SO and finally the fix:

```
sudo rm /etc/ssl/certs/java/cacerts
sudo update-ca-certificates -f
```

We should do this when we build `kedro-builder` image, but running it as part of a CI run works too for now.

### 🦠   Second bug: sbt (ivy2) and pytest-xdist's multi-processing don't nice together

This took _forever_ to find 😭  Below is an example log of the symptom:

```
:::: WARNINGS
		[FAILED     ] io.delta#delta-core_2.12;1.0.0!delta-core_2.12.jar: Downloaded file size doesn't match expected Content Length for https://repo1.maven.org/maven2/io/delta/delta-core_2.12/1.0.0/delta-core_2.12-1.0.0.jar. Please retry. (102ms)

		[FAILED     ] io.delta#delta-core_2.12;1.0.0!delta-core_2.12.jar: Downloaded file size doesn't match expected Content Length for https://repo1.maven.org/maven2/io/delta/delta-core_2.12/1.0.0/delta-core_2.12-1.0.0.jar. Please retry. (102ms)

	==== central: tried

	  https://repo1.maven.org/maven2/io/delta/delta-core_2.12/1.0.0/delta-core_2.12-1.0.0.jar

		[FAILED     ] org.antlr#antlr4;4.7!antlr4.jar: Downloaded file size doesn't match expected Content Length for https://repo1.maven.org/maven2/org/antlr/antlr4/4.7/antlr4-4.7.jar. Please retry. (48ms)

		[FAILED     ] org.antlr#antlr4;4.7!antlr4.jar: Downloaded file size doesn't match expected Content Length for https://repo1.maven.org/maven2/org/antlr/antlr4/4.7/antlr4-4.7.jar. Please retry. (48ms)

	==== central: tried

	  https://repo1.maven.org/maven2/org/antlr/antlr4/4.7/antlr4-4.7.jar

		[FAILED     ] org.antlr#ST4;4.0.8!ST4.jar: Downloaded file size doesn't match expected Content Length for https://repo1.maven.org/maven2/org/antlr/ST4/4.0.8/ST4-4.0.8.jar. Please retry. (12ms)

		[FAILED     ] org.antlr#ST4;4.0.8!ST4.jar: Downloaded file size doesn't match expected Content Length for https://repo1.maven.org/maven2/org/antlr/ST4/4.0.8/ST4-4.0.8.jar. Please retry. (12ms)

	==== central: tried

	  https://repo1.maven.org/maven2/org/antlr/ST4/4.0.8/ST4-4.0.8.jar

		::::::::::::::::::::::::::::::::::::::::::::::

		::              FAILED DOWNLOADS            ::

		:: ^ see resolution messages for details  ^ ::

		::::::::::::::::::::::::::::::::::::::::::::::

		:: io.delta#delta-core_2.12;1.0.0!delta-core_2.12.jar

		:: org.antlr#antlr4;4.7!antlr4.jar

		:: org.antlr#ST4;4.0.8!ST4.jar

		::::::::::::::::::::::::::::::::::::::::::::::
``` 

At first glance, it looks similar to the first problem, but it's not. It's a failed download, instead of an unresolved dependency. It says it can't download `org.antlr#antlr4;4.7!antlr4.jar` from any of the official repositories, i.e. maven, spark-apckages, etc., which makes no sense, because trying any of the link by itself, e.g. https://repo1.maven.org/maven2/org/antlr/antlr4/4.7/antlr4-4.7.jar, works just fine.

So what's going on here? I have a hunch after reading the line `Downloaded file size doesn't match expected Content Length` -- maybe two processes were trying to download these files concurrently and they interfere with each other, e.g. they all download to the same location and override each other's result. Not sure if that's actually the problem, but the hunch was correct. Running tests sequentially work. The error above only appears when running tests with pytest-xdist using more than 1 process.

The solution I implemented, which is adapted from pytest-xdist's own recommendation, is to just block all processes when a process is creating the spark session using a file lock. Works like a charm ✨ 🎩

### 💣.  Icing on the cake: OOM 

Just when I thought fixing the previous two problems were enough, [another cryptic error message](https://app.circleci.com/pipelines/github/quantumblacklabs/kedro/3668/workflows/4be1f64a-77a2-4650-9de0-c6a05e4b290c/jobs/70108) appeared:

```python
Traceback (most recent call last):
  File "/home/circleci/miniconda/envs/kedro_builder/lib/python3.8/site-packages/py4j/java_gateway.py", line 1207, in send_command
    raise Py4JNetworkError("Answer from Java side is empty")
py4j.protocol.Py4JNetworkError: Answer from Java side is empty
```

Fortunately, there is more in the error log:

```
self = <multiprocessing.popen_fork.Popen object at 0x7fb558b74b50>
process_obj = <ForkProcess name='ParallelRunnerManager-1' parent=2669 initial>

    def _launch(self, process_obj):
        code = 1
        parent_r, child_w = os.pipe()
        child_r, parent_w = os.pipe()
>       self.pid = os.fork()
E       OSError: [Errno 12] Cannot allocate memory
```

Right, it's an OOM (out of memory). This is where CircleCI's new Resource tab comes in very handy:

Screenshot 2021-12-03 at 01 49 47

We indeed maxed out on RAM, so I bumb the executors to `medium+`.

Now it's all passing 😆 

## Checklist

- [ ] Read the [contributing](https://github.com/quantumblacklabs/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/quantumblacklabs/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes

### CODE REVIEW

File: tests/extras/datasets/spark/conftest.py [hunk 1]
```diff
@@ -4,63 +4,38 @@
...
-
-# clean up pyspark after the test module finishes
-@pytest.fixture(scope="module")
-def spark_session():  # SKIP_IF_NO_SPARK
-    SparkSession.builder.getOrCreate = the_real_getOrCreate
-    builder = (
+def _setup_spark_session():
+    return configure_spark_with_delta_pip(
         SparkSession.builder.appName("MyApp")
+        .master("local[*]")
```
Comment (lorenabalan) [inline tests/extras/datasets/spark/conftest.py:20]: Is this telling Spark to run locally as opposed to a cluster, with however many cores it needs?

Comment (limdauto) [reply tests/extras/datasets/spark/conftest.py:20]: Yea https://stackoverflow.com/questions/32356143/what-does-setmaster-local-mean-in-spark -- but it wasn't needed. I was trying it to see if that's the issue. In the beginning I thought AWS env vars cause it to confuse to run elsewhere. No harm leaving it there though

**Review by MerelTheisenQB — COMMENTED** [review]
Wow great work @limdauto 👏 ⭐  I don't know too much about the spark test setup previously, but it does look simpler now. Thanks for the great write up as well, sounds like it was quite the debugging journey 🐛 👀

**Comment by lorenabalan** [comment]
THIS IS **GENIUS** THANK YOU SO SO MUCH!! ⭐ ⭐ ⭐ ⭐ ⭐

**Comment by datajoely** [comment]
You are incredible @limdauto

**Comment by mzjp2** [comment]
I loved reading this 😁
